src/zerialize.ts

```diff
@@ -320,92 +325,199 @@ const zerializers = {
+    opts.currentPath.pop();
+
+    opts.currentPath.push("value");
+    const value = s(def.valueType, opts);
+    opts.currentPath.pop();
+
+    return {
+      type: "record",
+      key,
+      value,
+    };
+  },
+  ZodMap: (def, opts) => {
+    opts.currentPath.push("key");
+    const key = s(def.keyType, opts);
+    opts.currentPath.pop();
+
+    opts.currentPath.push("value");
+    const value = s(def.valueType, opts);
+    opts.currentPath.pop();
+
+    return {
+      type: "map",
+      key,
+      value,
     };
   },
-  ZodRecord: (def, opts) => ({
-    type: "record",
-    key: s(def.keyType, opts),
-    value: s(def.valueType, opts),
-  }),
-  ZodMap: (def, opts) => ({
-    type: "map",
-    key: s(def.keyType, opts),
-    value: s(def.valueType, opts),
-  }),
 
   ZodEnum: (def) => ({ type: "enum", values: def.values }),
   // TODO: turn into enum
   ZodNativeEnum: () => ({ type: "unknown" }),
 
-  ZodUnion: (def, opts) => ({
-    type: "union",
-    options: def.options.map((opt) => s(opt, opts)),
-  }),
-  ZodDiscriminatedUnion: (def, opts) => ({
-    type: "discriminatedUnion",
-    discriminator: def.discriminator,
-    options: def.options.map((opt) => s(opt, opts)),
-  }),
-  ZodIntersection: (def, opts) => ({
-    type: "intersection",
-    left: s(def.left, opts),
-    right: s(def.right, opts),
-  }),
+  ZodUnion: (def, opts) => {
+    opts.currentPath.push("options");
 
-  ZodFunction: (def, opts) => ({
-    type: "function",
-    args: s(def.args, opts),
-    returns: s(def.returns, opts),
-  }),
-  ZodPromise: (def, opts) => ({ type: "promise", value: s(def.type, opts) }),
+    const union = {
+      type: "union",
+      options: def.options.map((opt, idx) => {
+        opts.currentPath.push(String(idx));
+        const result = s(opt, opts);
+        opts.currentPath.pop();
+        return result;
+      }),
+    };
+
+    opts.currentPath.pop();
+    return union;
+  },
+  ZodDiscriminatedUnion: (def, opts) => {
+    opts.currentPath.push("options");
+
+    const discriminatedUnion = {
+      type: "discriminatedUnion",
+      discriminator: def.discriminator,
+      options: def.options.map((opt, idx) => {
+        opts.currentPath.push(String(idx));
+        const result = s(opt, opts);
+        opts.currentPath.pop();
+        return result;
+      }),
+    };
+
+    opts.currentPath.pop();
+    return discriminatedUnion;
+  },
+  ZodIntersection: (def, opts) => {
+    opts.currentPath.push("left");
+    const left = s(def.left, opts);
+    opts.currentPath.pop();
+
+    opts.currentPath.push("right");
+    const right = s(def.right, opts);
+    opts.currentPath.pop();
+
+    return {
+      type: "intersection",
+      left,
+      right,
+    };
+  },
+
+  ZodFunction: (def, opts) => {
+    opts.currentPath.push("args");
+    const args = s(def.args, opts);
+    opts.currentPath.pop();
+
+    opts.currentPath.push("returns");
+    const returns = s(def.returns, opts);
+    opts.currentPath.pop();
```

Comment:
I've dropped the mutative style. I think the former style worked because it was meticulous in reverting the changes and each step would eventually return, but in any case this works also.

Reply:
> I've dropped the mutative style. I think the former style worked because it was meticulous in reverting the changes and each step would eventually return, but in any case this works also.

It did look meticulous in that way, I did not mean to doubt that.

Yet the reference to the array that was passed in, is also the one that was mutated thereafter. So if anything with the `s=>zerializeRefs` call were to use or mutate the ref at a later point, wouldn't this cause the spooky distance action that I fear?

> I also added a commit to revert back to the terser style of using arrow functions without return statements. I personally find that style cumbersome in debugging and refactoring, but I added the commit if you really do prefer it.

While I do agree that debugging takes more steps, I prefer optimizing for reading on code that is straightforward enough where I don't expect a ton of debugging sessions. So it's a bit of a subjective value judgment.
Though even if, thankfully VSCode added some helpful structural editing commands here: when your cursor is within a terse arrow fn, press `Cmd + .` (aka "Quick Fix") and hit `Enter` (on "Add braces to arrow function"). So it's only 3 keypresses to make it more debuggable which I find to be an okay cost

Reply:
Re: arrow functions, sweet, thanks for the tip... Re: the mutating array, I'll admit it's hard to get my head around it. And thanks for the review!